extensions-compose-jetbrains/src/commonMain/kotlin/com/arkivanov/decompose/extensions/compose/jetbrains/pages/Pages.kt

```diff
@@ -52,10 +53,13 @@ fun <T : Any> Pages(
     modifier: Modifier = Modifier,
     scrollAnimation: PagesScrollAnimation = PagesScrollAnimation.Disabled,
     pager: Pager = defaultHorizontalPager(),
-    pageContent: @Composable (index: Int, page: T) -> Unit,
+    pageContent: @Composable PagerScope.(index: Int, page: T) -> Unit,
 ) {
     val selectedIndex = pages.selectedIndex
-    val state = rememberPagerState(initialPage = selectedIndex)
+    val state = rememberPagerState(
+        initialPage = selectedIndex,
+        pageCount = { pages.items.size },
```

Comment:
After reviewing the code further, it looks like the remembered `pageCount` lambda captures the instance of `ChildPages`, which is immutable. So even if a new instance of `ChildPages` of another size will be be supplied later, the lambda will be still capturing the old one.

It feels that this overload doesn't make any sense now, and we can only have the one with `Value<ChildPages>`. WDYT?

Reply:
Or perhaps even better, let the second overload accept `State<ChildPages>` instead of just `ChildPages`. And make sure that `pageCount` lambda reads the actual value from the provided `State`.

```kotlin
@Composable
fun <T : Any> Pages(
    pages: State<ChildPages<*, T>>,
    onPageSelected: (index: Int) -> Unit,
    modifier: Modifier = Modifier,
    scrollAnimation: PagesScrollAnimation = PagesScrollAnimation.Disabled,
    pager: Pager = defaultHorizontalPager(),
    pageContent: @Composable PagerScope.(index: Int, page: T) -> Unit,
)
```

What's your opinion on this?

Reply:
Now that I have an updated and working version I can see the problem too. I think your suggested alternative with `State<ChildPages<*, T>>` would be more flexible for the users, since they have more options for using `Pages` (e.g. when using `StateFlow` instead of `Value`).

Should I update the second function so that it consumes the state passed via `pages` and use the newly consumed state instead? And is it an issue if consuming twice the state in the case of `Pages` with `pages: Value`?

Reply:
In addition to that, can we create a few tests for the updating part of `ChildPages`? Because I am experiencing a confusing `IndexOutOfBoundsException` when I simply use a mutable state property and update the pages from two to three pages.

I was trying to add the following test for the variant `pages: State`. I feel like I am missing something crucial:

```kotlin
@OptIn(ExperimentalFoundationApi::class, ExperimentalDecomposeApi::class)
@Test
fun GIVEN_pages_dislpayed_WHEN_pages_state_updated_THEN_pageCount_updated() {
    runBlocking(Dispatchers.Main) {
        val state = mutableStateOf(
            ChildPages(
                items = listOf(
                    Child.Created(Config.Config1, Unit),
                    Child.Destroyed(Config.Config2),
                ),
                selectedIndex = 0,
            ),
        )

        composeRule.setContent {
            Pages(
                pages = state,
                onPageSelected = {},
            ) { index, page ->
                Text("$index $page")
            }
        }

        composeRule.awaitIdle()
        state.value = ChildPages(
            items = listOf(
                Child.Destroyed(Config.Config1),
                Child.Created(Config.Config2, Unit),
                Child.Destroyed(Config.Config3), // <-- Tried both, created and destroyed
            ),
            selectedIndex = 1,
        )
        composeRule.waitForIdle() // <-- Fails here with IndexOutOfBoundsException
    }
}

private sealed class Config {

    object Config1: Config()

    object Config2: Config()

    object Config3: Config()
}
```

Reply:
Thanks! Tests are always welcome!

I have reproduced that crash without Decompose, filed https://issuetracker.google.com/issues/295653367 just in case. Let's how it goes.

I believe that's a Compose bug, but here is a code that seems to work.

```kotlin
@ExperimentalFoundationApi
@ExperimentalDecomposeApi
@Composable
fun <T : Any> Pages(
    pages: State<ChildPages<*, T>>,
    onPageSelected: (index: Int) -> Unit,
    modifier: Modifier = Modifier,
    scrollAnimation: PagesScrollAnimation = PagesScrollAnimation.Disabled,
    pager: Pager = defaultHorizontalPager(),
    pageContent: @Composable PagerScope.(index: Int, page: T) -> Unit,
) {
    val selectedIndex = pages.value.selectedIndex
    val state = rememberPagerState(
        initialPage = selectedIndex,
        pageCount = { pages.value.items.size },
    )

    LaunchedEffect(selectedIndex) {
        if (state.currentPage != selectedIndex) {
            when (scrollAnimation) {
                is PagesScrollAnimation.Disabled -> state.scrollToPage(selectedIndex)
                is PagesScrollAnimation.Default -> state.animateScrollToPage(page = selectedIndex)
                is PagesScrollAnimation.Custom -> state.animateScrollToPage(page = selectedIndex, animationSpec = scrollAnimation.spec)
            }
        }
    }

    DisposableEffect(state.settledPage) {
        onPageSelected(state.settledPage)
        onDispose {}
    }

    pager(
        modifier,
        state,
        { pages.value.items[it].configuration },
    ) { pageIndex ->
        pages.value.items[pageIndex].instance?.also { page ->
            pageContent(pageIndex, page)
        }
    }
}
```

What do you think?

Reply:
Seems like when the `pager` receives an immutable value through the function it doesn't update, like with the `rememberPagerState.pageCount` for which we introduced the `State<ChildPages<*, T>>`.

The solution you provided works now with the test too. I will Add a commit with the updates later today and look into adding a few tests afterwards. :)

Reply:
While adding tests I noticed that they are not in `commonTest`. Is there a specific reason for that or can I add the `PagesTests` to `commonTest`?

Reply:
Tests are in `jvmMain` because they depend on Compose for Desktop testing stuff. I think we can't run common tests for Compose yet?

Reply:
Seems like tests were not committed. Do you still want to add the tests?

Reply:
I haven't committed them yet because I wasn't sure if I should add them into `common` code. But if they are not yet supported I will have to add them for `jvm` only.

Reply:
`jvmTest` seems like a good place for now. Thanks!